### ISSUE TITLE
Adding Universal Language Model Fine-tuning ULMFiT pre-trained LM to spacy and alowing a simple way to train new models

### ISSUE THREAD

**Issue by jmizgajski** [explosion/spaCy#2342]
## Feature description

[Universal Language Model Fine-tuning for Text Classification](https://arxiv.org/pdf/1801.06146.pdf) presented a novel method to fine tune a pre-trained universal  language model to a particular classification task which achieved beyond state-of-the art (18-24% reduction in error rate) on multiple benchmark text classification tasks. The fine tuning requires *very* few examples (100) to achieve very good results. 

Here is an excerpt of the abstract which provides a good TL;DR of the paper (duh):

Inductive transfer learning has greatly impacted
computer vision, but existing approaches
in NLP still require task-specific
modifications and training from scratch.
We propose Universal Language Model
Fine-tuning (ULMFiT), an **effective transfer
learning method** that can be applied to
any task in NLP, and introduce techniques
that are key for fine-tuning a language
model. Our method significantly **outperforms
the state-of-the-art on six text classification
tasks, reducing the error by 18-
24%** on the majority of datasets. Furthermore,
with **only 100 labeled examples**, it
**matches the performance** of training from
scratch on **100× more data**. We opensource
our pretrained models and code

I propose that spacy adds their pre-trained models and a simple way to fine tune to a new task as a core feature of the library. 

## Could the feature be a [custom component](https://spacy.io/usage/processing-pipelines#custom-components) or [spaCy plugin](https://spacy.io/universe)?
If so, we will tag it as [`project idea`](https://github.com/explosion/spaCy/labels/project%20idea) so other users can take it on.

This seems like a core feature of spacy, greatly increasing its industrial potential. I would argue to make it a first class citizen if authors and licensing of this work permits that.

**Comment by wojciak** [comment]
This! 👍

**Comment by ngoel17** [comment]
Like It.

**Comment by sebastianruder** [comment]
Author here. I'd love to see this happen and I'm sure @jph00 would also be on board. Fast.ai is working on [pre-trained models for other languages](http://forums.fast.ai/t/language-model-zoo-gorilla/14623) and we'll be working to simplify and make the code more robust.

**Comment by jph00** [comment]
For sure - I discussed the basic idea of LM fine-tuning with @honnibal recently. I'd be happy to improve integration between fastai's language modeling and forthcoming model zoo and spacy. Our model zoo should work fine with anything based on pytorch - to work with thinc would require porting the architecture and weights, of course.

**Comment by DuyguA** [comment]
Would love to do the pre-trained Turkish model!

**Comment by honnibal** [comment]
Super keen on this! @jph00 the vision for plugging in other libraries is to have Thinc as a thin wrapper on top of other libraries. There's already a PR for exposing PyTorch models as Thinc models that I need to review: https://github.com/explosion/thinc/pull/61 . There's no reason to use Thinc as the runtime.

One experiment I'd love to try is using the LM output to supervise the current CNN architecture I've been using. The CNN is very fast on CPU, but I haven't found a good way to pre-train it yet. If we use your LM as the supervision, we have unlimited training data...

**Comment by slavakurilyak** [comment]
Going from initializing the first layer of our models to pretraining the entire model with hierarchical representations is a must! For additional inspiration, check out [NLP's ImageNet moment has arrived](https://thegradient.pub/nlp-imagenet/) by The Gradient.

**Comment by ines** [comment]
See the latest nightly releases: https://github.com/explosion/spaCy/releases/tag/v2.1.0a3

**Comment by superzadeh** [comment]
Amazing work @ines @honnibal and all the other contributors, can't wait to give this a shot!
